### PR TITLE
Commendations fix

### PR DESCRIPTION
**Pull request by Shoes01**
I added one include in order to be able to compile. Something to do with using yaml-cpp 0.5.3.

Converting _killCriteria from a map to a pair is the primary fix, and having now functional code exposed another hole, which was fixed with the addition of the bool andCriteriaMet.

A sanity check was added at line 377 of SoldierDiary.cpp

### CODE REVIEW

File: src/Savegame/SoldierDiary.cpp [hunk 1]
```diff
@@ -422,17 +428,15 @@ bool SoldierDiary::manageCommendations(Mod *mod, std::vector<MissionStatistics*>
...
-							bool foundMatch = true;
 							
-							// Loop over the DETAILs of the AND vector.
+							// Loop over the DETAILs of one AND vector.
 							for (std::vector<std::string>::const_iterator detail = andCriteria->second.begin(); detail != andCriteria->second.end(); ++detail)
-							{
-								std::string battleTypeArray[] = { "BT_NONE", "BT_FIREARM", "BT_AMMO", "BT_MELEE", "BT_GRENADE",
-									"BT_PROXIMITYGRENADE", "BT_MEDIKIT", "BT_SCANNER", "BT_MINDPROBE", "BT_PSIAMP", "BT_FLARE", "BT_CORPSE", "BT_END" };
+							{							
+								std::string battleTypeArray[] = { "BT_NONE", "BT_FIREARM", "BT_AMMO", "BT_MELEE", "BT_GRENADE",	"BT_PROXIMITYGRENADE", "BT_MEDIKIT", "BT_SCANNER", "BT_MINDPROBE", "BT_PSIAMP", "BT_FLARE", "BT_CORPSE", "BT_END" };
```
Comment (MeridianOXC) [inline src/Savegame/SoldierDiary.cpp]: Btw. this should go way more up... why construct a new array each time within a loop? Can easily be a constant on class level.

File: src/Savegame/SoldierDiary.cpp [hunk 2]
```diff
@@ -442,8 +446,7 @@ bool SoldierDiary::manageCommendations(Mod *mod, std::vector<MissionStatistics*>
 									}
 								}
 
-								std::string damageTypeArray[] = { "DT_NONE", "DT_AP", "DT_IN", "DT_HE", "DT_LASER", "DT_PLASMA",
-									"DT_STUN", "DT_MELEE", "DT_ACID", "DT_SMOKE", "DT_END"};
+								std::string damageTypeArray[] = { "DT_NONE", "DT_AP", "DT_IN", "DT_HE", "DT_LASER", "DT_PLASMA", "DT_STUN", "DT_MELEE", "DT_ACID", "DT_SMOKE", "DT_END"};
```
Comment (MeridianOXC) [inline src/Savegame/SoldierDiary.cpp]: Same here.

File: src/Savegame/SoldierDiary.cpp [hunk 3]
```diff
@@ -442,8 +446,7 @@ bool SoldierDiary::manageCommendations(Mod *mod, std::vector<MissionStatistics*>
...
 								int damageType = 0;
 								for (; damageType != 11; ++damageType)
```
Comment (MeridianOXC) [inline src/Savegame/SoldierDiary.cpp]: Constant instead of magic number 11?

File: src/Savegame/SoldierDiary.cpp [hunk 4]
```diff
@@ -422,17 +428,15 @@ bool SoldierDiary::manageCommendations(Mod *mod, std::vector<MissionStatistics*>
...
-							// Loop over the DETAILs of the AND vector.
+							// Loop over the DETAILs of one AND vector.
 							for (std::vector<std::string>::const_iterator detail = andCriteria->second.begin(); detail != andCriteria->second.end(); ++detail)
-							{
-								std::string battleTypeArray[] = { "BT_NONE", "BT_FIREARM", "BT_AMMO", "BT_MELEE", "BT_GRENADE",
-									"BT_PROXIMITYGRENADE", "BT_MEDIKIT", "BT_SCANNER", "BT_MINDPROBE", "BT_PSIAMP", "BT_FLARE", "BT_CORPSE", "BT_END" };
+							{							
+								std::string battleTypeArray[] = { "BT_NONE", "BT_FIREARM", "BT_AMMO", "BT_MELEE", "BT_GRENADE",	"BT_PROXIMITYGRENADE", "BT_MEDIKIT", "BT_SCANNER", "BT_MINDPROBE", "BT_PSIAMP", "BT_FLARE", "BT_CORPSE", "BT_END" };
 								int battleType = 0;
 								for (; battleType != 13; ++battleType)
```
Comment (MeridianOXC) [inline src/Savegame/SoldierDiary.cpp]: Constant instead of magic number 13, also good to use that constant as array size

Comment (Shoes01) [inline src/Savegame/SoldierDiary.cpp]: Thanks for the input! I can't test from here, but I will likely push another commit today that conforms to your array size idea.

**Comment by Shoes01** [comment]
The latest commit fixes a few things. Commendations can be awarded even if the base is lost; I fixed an error where succession kills were being forgotten by the time commendation awards were decided; weapon and grenade kills no longer count friendly fire.

**Comment by Shoes01** [comment]
Latest commit opens the "Best of" series of commendations to have criteria based on score, rather than being hard coded. As such, the "Best Squaddie" could need a minimum score of 100, whereas the "Best Overall" could need a minimum score of 2000.

**Comment by MeridianOXC** [comment]
What's needed for this to be merged?

**Comment by SupSuper** [comment]
@MeridianOXC I was waiting for @Shoes01 to tell me when he's done, you never know with him. :P

**Comment by Shoes01** [comment]
The burst of motivation I had was redirected to a different project IRL, so I won't be adding to this PR. 

:D

**Pull request merged**